### ISSUE TITLE
adapt() method of a CategoryEncoding layer returns the wrong shape when fitted to Integer list not starting at 0

### ISSUE THREAD

**Issue by scd75** [tensorflow/tensorflow#42680]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.0-dev20200825
- Python version: 3.7.8

**Describe the current behavior**
When calling "adapt()" method on a CategoryEncoding layer with either:
- a dataset feature with integers not starting at 0 (eg. month identifiers from 1 to 12)
- a np.array with a list of possible integers not starting at 0
the resulting shape of the CategoryEncoding layer seems to be determined by the max integer, and not the cardinality of integers in the list / data.
eg, we get a shape of 13 if we adapt the layer to an integer list ranging from 1 to 12, or even from 3 to 12 for example

**Describe the expected behavior**
Expect the adapt method of a CategoryEncoding layer to output an object with a shape equal to the number of unique values in the integer-coded feature

**Standalone code to reproduce the issue**
```

import numpy as np
import tensorflow as tf

month_input_layer = tf.keras.layers.Input(shape=(1,), name="month", dtype="int64")
month_encoded_layer = tf.keras.layers.experimental.preprocessing.CategoryEncoding(output_mode="binary")
dense_layer = tf.keras.layers.Dense(1,activation="relu")

month_encoded_layer.adapt(np.arange(1,13))
output = dense_layer(month_encoded_layer(month_input_layer))

model = tf.keras.models.Model(month_input_layer, output)

model.summary()
```

**Other info / logs**:
```

2020-10-05 10:12:27.032788: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-10-05 10:12:27.103511: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x1dbad9023f0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-10-05 10:12:27.112275: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
(.venv) PS C:\Users\SamuelCheptou\CSN Energy\Marché - Paris Restricted - General\5# Ressources & Données\CDC\Analyses CDC forecasting\scripts python SC>  cd 'c:\Users\SamuelCheptou\CSN Energy\Marché - Paris Restricted - General\5# Ressources & Données\CDC\Analyses CDC forecasting\scripts python SC'; & 'c:\Users\SamuelCheptou\local_CDC\.venv\Scripts\python.exe' 'c:\Users\SamuelCheptou\.vscode\extensions\ms-python.python-2020.9.112786\pythonFiles\lib\python\debugpy\launcher' '57146' '--' 'c:\Users\SamuelCheptou\CSN Energy\Marché - Paris Restricted 
- General\5# Ressources & Données\CDC\Analyses CDC forecasting\scripts python SC\Github_bug.py' 
2020-10-05 10:31:37.812387: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-10-05 10:31:37.833441: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x241a1d351f0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-10-05 10:31:37.861026: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================
month (InputLayer)           [(None, 1)]               0
_________________________________________________________________
category_encoding (CategoryE (None, 13)                1
_________________________________________________________________
dense (Dense)                (None, 1)                 14
=================================================================
Total params: 15
Trainable params: 14
Non-trainable params: 1
_________________________________________________________________

```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/78ee390cd9917f7a73037611f844d214/42680.ipynb). Thanks!

**Comment by scd75** [comment]
Hello, any update on this issue ? thanks!

**Comment by Saduf2019** [comment]
@scd75 
I ran the code on nightly '2.5.0-dev20210114', please refer to the [gist here](https://colab.research.google.com/gist/Saduf2019/c88dd3d964b7712420ce4e64acc26191/untitled499.ipynb)

**Comment by scd75** [comment]
can somebody take a look at this issue ?
It has now been reproduced 2 times, but no solution was given so far.. Thanks!

**Comment by Saduf2019** [comment]
@scd75 
Please try on latest version [nightly[ and let us know if you still face the issue.

**Comment by scd75** [comment]
Yes I just tried and still face the issue on the last nightly. The shape of the "category_encoding" layer should be (None, 12) instead of (None, 13)

**Comment by ymodak** [comment]
@scd75 Apologies for the delay here. If this is still an issue; TF 2.5 was recently released. I am not able to verify the fix since the code snippet is incomplete. Perhaps you can share complete reproducible code to test.

**Comment by bhack** [comment]
You could verify in TF 2.5 but 2.5 but:

@scd75  So you need   specify the `num_tokens` arg in `tf.keras.layers.experimental.preprocessing.CategoryEncoding`

OR

> For integer inputs where the total number of tokens is not known, see tf.keras.layers.experimental.preprocessing.IntegerLookup

**Comment by ymodak** [comment]
TF 2.5 raises `ValueError: num_tokens must be set to use this layer. If the number of tokens is not known beforehand, use the IntegerLookup layer instead.`

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.5.1 (v2.5.0-160-g8222c1cfc86).

- `(tf.keras.layers.LayerNormalization)`, `tf.compat.v1.keras.layers.LayerNormalization`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.keras.layers.LayerNormalization</code> in older versions</summary>

```python
import numpy as np
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


month_input_layer = tf.keras.layers.Input(shape=(1,), name="month", dtype="int64")
month_encoded_layer = tf.compat.v1.keras.layers.experimental.preprocessing.CategoryEncoding(output_mode="binary")
dense_layer = tf.keras.layers.Dense(1,activation="relu")

month_encoded_layer.adapt(np.arange(1,13))
output = dense_layer(month_encoded_layer(month_input_layer))

model = tf.keras.models.Model(month_input_layer, output)

model.summary()
```

On my GPU machine, the model summary is printed without any Exceptions, which is unexpected.

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
month (InputLayer)           [(None, 1)]               0         
_________________________________________________________________
category_encoding (CategoryE (None, 13)                0         
_________________________________________________________________
dense (Dense)                (None, 1)                 14        
=================================================================
Total params: 14
Trainable params: 14
Non-trainable params: 0
_________________________________________________________________
```

This behavior is also reproducible on my CPU machine:

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
month (InputLayer)           [(None, 1)]               0         
_________________________________________________________________
category_encoding (CategoryE (None, 13)                0         
_________________________________________________________________
dense (Dense)                (None, 1)                 14        
=================================================================
Total params: 14
Trainable params: 14
Non-trainable params: 0
_________________________________________________________________
```
</details>

It seems to be fixed in tensorflow 2.6.0 (v2.6.0-rc2-32-g919f693420e) and later versions.

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.4.3 (v2.4.2-142-g72bb4c22adb).

- `(tf.keras.layers.experimental.preprocessing.CategoryEncoding)`, `tf.compat.v1.keras.layers.experimental.preprocessing.CategoryEncoding`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.keras.layers.experimental.preprocessing.CategoryEncoding</code> in older versions</summary>

```python
import numpy as np
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


month_input_layer = tf.keras.layers.Input(shape=(1,), name="month", dtype="int64")
month_encoded_layer = tf.compat.v1.keras.layers.experimental.preprocessing.CategoryEncoding(output_mode="binary")
dense_layer = tf.keras.layers.Dense(1,activation="relu")

month_encoded_layer.adapt(np.arange(1,13))
output = dense_layer(month_encoded_layer(month_input_layer))

model = tf.keras.models.Model(month_input_layer, output)

model.summary()
```

On my GPU machine, the model summary is printed without any Exceptions, which is unexpected.

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
month (InputLayer)           [(None, 1)]               0         
_________________________________________________________________
category_encoding (CategoryE (None, 13)                0         
_________________________________________________________________
dense (Dense)                (None, 1)                 14        
=================================================================
Total params: 14
Trainable params: 14
Non-trainable params: 0
_________________________________________________________________
```

This behavior is also reproducible on my CPU machine:

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
month (InputLayer)           [(None, 1)]               0         
_________________________________________________________________
category_encoding (CategoryE (None, 13)                0         
_________________________________________________________________
dense (Dense)                (None, 1)                 14        
=================================================================
Total params: 14
Trainable params: 14
Non-trainable params: 0
_________________________________________________________________
```
</details>

It seems to be fixed in tensorflow 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) and later versions.
It throws a `ValueError` Exception and tells users that `num_tokens must be set to use this layer, and if the number of tokens is not known beforehand, use the IntegerLookup layer instead`.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
